### PR TITLE
feat(whitelist-config): add rule config to whitelist certain functions

### LINKED ISSUE #73: Config to assume upstream functions are pure
This plugin is pretty much exactly what I'm looking for to prevent side effects! Thanks for all the work that went into it. There is, however, one thing though is preventing me from using this for my projects, and that's how many things you have to manually mark using magic comments. It makes perfect sense as to why, but it makes this plugin infeasible to use for medium to larger projects with lots of modules that import from both each other and 3rd party libraries.

I think the thing that would help most here would be the ability to mark functions as side effect free in the plugin config. This would really bring down the number of places you need to explicitly mark `no-side-effects-when-called` in code, making this plugin much more approachable for medium to larger projects!

Here's what it could possibly look like in config:

```json
{
  "rules": {
    "tree-shaking/no-side-effects-in-initialization": ["error", {
      "noSideEffectsWhenCalled": [
        {"module": "react", "functions": ["memo", "createComponent"]},
        {"module": "styled-components", "functions": "*"},
        {"builtin": "Object.values"},
        {"module": "#local", "functions": ["myFunc1", "myFunc2"]}
      ]
    }]
  }
}
```

I hope this example is mostly self explanatory, but to call out some extra stuff I added for convenience:
* You can white list all functions for a given module using `"*"` instead of passing an array of function names
* You can white list functions from all "local" modules by using the special module name `"#local"`. What I mean by a local module is one that is imported via a relative path.

This would also help address #43 by allowing folks to explicitly whitelist things themselves. Specifically since someone may be using a bundler besides rollup that already correctly handles a given builtin.

### PR DESCRIPTION
**Pull request by maclockard**
I ended up going ahead and implementing #73. I can add some tests if we decide to move forward with this PR. Let me know what you think!

fixes #73

### CODE REVIEW

File: lib/utils/helpers.js [hunk 1]
```diff
@@ -26,6 +26,7 @@ const flattenMemberExpressionIfPossible = node => {
       if (node.computed || node.property.type !== 'Identifier') {
         return null
       }
+      // eslint-disable-next-line no-case-declarations
       const flattenedParent = flattenMemberExpressionIfPossible(node.object)
```
Comment (maclockard) [inline lib/utils/helpers.js:30]: this lint was failing lint check when I committed even though I didn't touch it, let me know what you want me to do

File: lib/utils/helpers.js [hunk 2]
```diff
@@ -57,13 +67,44 @@ const getTreeShakingComments = comments => {
   return { has: token => treeShakingComments.indexOf(token) >= 0 }
 }
 
+const isFunctionSideEfectFree = (functionName, moduleName, contextOptions) => {
```
Comment (lukastaegert) [inline lib/utils/helpers.js:70]: Typo: `isFunctionSideEffectFree`

**Comment by wSedlacek** [comment]
@maclockard Do you still plan to work on this or is it safe for someone else to continue from where you left off?

**Comment by lukastaegert** [comment]
Unfortunately at the moment I find little time to work on this plugin as most of my extra time is spent on Rollup. Any outside help is welcome.

**Comment by maclockard** [comment]
I do plan to work on it. I've been using a personal fork with this rule heavily at work, I just need to get around adding some tests. will try to do it in the next 2ish weeks

**Pull request merged**